docs/design/endpoints.md

```diff
@@ -63,52 +63,231 @@ Client config will include a field for the resolver unique to that service, and
-own, but rather exposes an interface in the middleware generator for SDK authors to define and resolve their own
-builtins.
+Annotates a field in a structure which has been designated as an operation's input. Designates a field on that operation
+to be bound to a provider parameter at resolution time.
+
+If a context param targets a value previously declared as a client context param, the per-operation value overrides the
+one configured by the client for that call.
 
-For example, the AWS SDK has special builtins for the AWS region, as well as for concepts unique to the AWS ecosystem
-such as the usage of dual-stack endpoints. The AWS-specific generator implementation over top of the base smithy-kotlin
-one is responsible for providing the logic to code-generate the sourcing of these builtins in the resolver middleware
-(detailed below).
+For example, given the following parameters/model snippet:
+
+```json
+{
+  "regionPrefix": {
+    "type": "string"
+  }
+}
+```
+
+```smithy
+@clientContextParams(
+    regionPrefix: {type: "string", documentation: "the country code for the client region"}
+)
+service BasicService {
+    operations: [
+        GetItem,
+        GetSecretItem
+    ]
+}
+
+operation GetItem {
+    input: GetItemInput;
+}
+
+structure GetItemInput {
+    @contextParam(name: "regionPrefix")
+    regionPrefix: string;
+}
+
+operation GetSecretItem {
+    input: GetSecretItemInput;
+}
+
+structure GetSecretItemInput { }
+```
+
+The value of `regionPrefix` would use the request value for `GetItem`, whereas `GetSecretItem` would bind from the
+client config.
+
+#### Built-ins
+
+The rule set specification allows for the definition of built-ins, which are SDK-specific values that are bound to
+provider parameters at the time of resolution. As a generic codegen provider, smithy-kotlin doesn't implement any
+built-ins on its own but rather exposes an interface in the middleware generator for SDK authors to define and resolve
+their own built-ins.
+
+For example, the AWS SDK has a special built-in for the AWS region. Given the following parameter declaration:
+
+```json
+{
+  "region": {
+    "type": "string",
+    "builtin": "AWS::Region"
+  }
+}
+```
+
+The base middleware generator cannot render the binding of this value. The AWS SDK must implement custom logic to bind
+this value from the client config.
 
 ## Codegen
+### Package "endpoints"
+The provider parameters and interface will generate to a new top-level `endpoints` package.
+
+### Provider parameters
+
+Provider parameters will be generated as a data class, since backwards compatability of parameter values is guaranteed (
+field order MUST be preserved, and obsolete values are marked deprecated and not removed).
+
+The presence of required values is checked at construction, with an error thrown if any are missing.
+
+The generation includes DSL builder support for several reasons:
+1. caller convenience - SDK users are able to implement and call their own endpoint providers
```

Comment:
> and call their own endpoint providers

Not sure I understand this part. Why would calling an endpoint provider be an argument for builders?